### PR TITLE
Initial pull request

### PR DESCRIPTION
**Pull request by zschira**
Pull request addressing some initial [comments](https://github.com/catalyst-cooperative/xbrl-extract/issues/1). This pull request improves comments and documentation, logging, fixes some command line arguments, migrates to using a process pool instead of a thread pool. The thread pool was providing some minor performance improvements on my machine, but the process pool does seem to do a better job speeding up parsing if the `--batch-size` is large enough, otherwise the overhead of passing data between processes outweighs any improvements. The benefits of the process pool also become more pronounced as more filings are processed.

### CODE REVIEW

**Review by zaneselvans — COMMENTED** [review]
I haven't gotten through all of this and have a backed up list of PRs I'm already on, so I'm going to submit as it is and tag @bendnorman -- especially since there's not as much overlap (sadly) with our old FERC work as we might have hoped.

File: src/xbrl_extract/arelle_interface.py [hunk 1]
```diff
@@ -15,19 +15,40 @@ def load_xbrl(path: str):
     return ModelXbrl.load(model_manager, path)
 
 
+def load_taxonomy(path: str):
```
Comment (bendnorman) [inline src/xbrl_extract/arelle_interface.py:18]: Can you include argument descriptions in the docstrings?

File: src/xbrl_extract/cli.py [hunk 2]
```diff
@@ -14,83 +18,87 @@ def parse_main():
...
+    logger.setLevel(args.loglevel)
+    log_format = "%(asctime)s [%(levelname)8s] %(name)s:%(lineno)s %(message)s"
+    coloredlogs.install(fmt=log_format, level=args.loglevel, logger=logger)
+
+    if args.logfile:
+        file_logger = logging.FileHandler(args.logfile)
+        file_logger.setFormatter(logging.Formatter(log_format))
+        logger.addHandler(file_logger)
+
+    engine = create_engine(f"sqlite:///{args.sql_path}")
```
Comment (bendnorman) [inline src/xbrl_extract/cli.py:98]: Later on you write the dataframes to the database using `pd.DataFrame.to_sql()`. [SQLAlchemy recommends](https://docs.sqlalchemy.org/en/14/core/connections.html#sqlalchemy.engine.Engine) using the python context manager so the database connection is closed once the dataframe is written.

File: src/xbrl_extract/cli.py [hunk 3]
```diff
@@ -14,83 +18,87 @@ def parse_main():
...
+
+    if args.clobber:
+        helpers.drop_tables(engine)
```
Comment (bendnorman) [inline src/xbrl_extract/cli.py:101]: Do you create a table schema for the data?

File: src/xbrl_extract/helpers.py [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+    """
+    logger = logging.getLogger(__name__)
+    logger.info("Dropping tables")
+
+    md = sa.MetaData()
+    md.reflect(engine)
+    md.drop_all(engine)
+    conn = engine.connect()
+    conn.exec_driver_sql("VACUUM")
+    conn.close()
```
Comment (bendnorman) [inline src/xbrl_extract/helpers.py:33]: You can use the context manager here instead of calling `con.close()`.

**Review by bendnorman — CHANGES_REQUESTED** [review]
Looks pretty good to me. Is there a high-level overview of how XBRL and the extraction work?

File: src/xbrl_extract/cli.py [hunk 5]
```diff
@@ -43,23 +41,22 @@ def parse_main():
...
-        help="Specify number of threads in pool (defaults to `os.cpu_count()`)"
-    )
-    parser.add_argument(
-        "--verbose",
-        action="store_true",
-        default=False,
-        help="Print logging messages to stdout",
+        help="Specify number of workers in pool (will attempt to choose a reasonable default if not specified)"
     )
     parser.add_argument(
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/cli.py:52]: Added

File: src/xbrl_extract/__init__.py [hunk 6]
```diff
@@ -0,0 +1,3 @@
+import logging
+
+logging.getLogger(__name__).addHandler(logging.NullHandler())
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/__init__.py:5]: I was copying `PUDL` with this, but I'm not sure it's necessary. I feel like the docs don't really provide much info on best practices. Maybe there's a good blog or something we could use for guidance?

File: src/xbrl_extract/cli.py [hunk 7]
```diff
@@ -78,19 +75,32 @@ def get_instances(instance_path: Path):
             raise ValueError("Must provide valid path to XBRL instance or "
                              "directory of XBRL instances.")
 
-        return [(str(instance_path), 0)]
+        return [(str(instance_path), instance_path.name.rstrip(instance_path.prefix))]
     # Directory of instances
     elif instance_path.is_dir():
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/cli.py]: Updated. It should now raise an error if pointed to anything other than a file or directory

File: src/xbrl_extract/cli.py [hunk 8]
```diff
@@ -78,19 +75,32 @@ def get_instances(instance_path: Path):
...
-        instances = [str(instance) for instance in instance_path.iterdir()
-                     if instance.suffix in ALLOWABLE_SUFFIXES]
-        return [(instance, i) for i, instance in enumerate(sorted(instances))]
+        return [(str(instance), instance.name.rstrip(instance.suffix))
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/cli.py]: Added return type

File: src/xbrl_extract/helpers.py [hunk 9]
```diff
@@ -0,0 +1,31 @@
+import sqlalchemy as sa
+import logging
+
+
+def drop_tables(engine):
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/helpers.py]: Updated

File: src/xbrl_extract/cli.py (hunk 2 above)
Comment (zschira) [reply src/xbrl_extract/cli.py:114]: Fixed

File: src/xbrl_extract/arelle_interface.py (hunk 1 above)
Comment (zschira) [reply src/xbrl_extract/arelle_interface.py:23]: Fixed

File: src/xbrl_extract/helpers.py (hunk 4 above)
Comment (zschira) [reply src/xbrl_extract/helpers.py]: Fixed

File: src/xbrl_extract/cli.py (hunk 3 above)
Comment (zschira) [reply src/xbrl_extract/cli.py:117]: Previously the schema was sort of defined implicitly. With the update to generate a frictionless datapackage descriptor, it now explicitly creates a schema which is used to initialize the dataframes as they're created

File: src/xbrl_extract/instance.py [hunk 10]
```diff
@@ -152,19 +152,17 @@ def in_axes(self, axes: List[str]) -> bool:
         """
         return self.entity.in_axes(axes)
 
-    def get_context_ids(self, filing_id: int = None) -> Dict:
+    def get_context_ids(self, filing_name: str) -> Dict:
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/instance.py:186]: I've included the filing name (which is just the name of the xbrl file) as way to try and guarantee that the primary key for each table will uniquely identify records. I believe it's theoretically possible that you could end up with XBRL facts from the same filer with the same report period in two separate filings. I don't think this should happen under normal circumstances, but it's possible that a filer could have their dates screwed up in a way that could cause this sort of collision

File: src/xbrl_extract/instance.py [hunk 11]
```diff
@@ -152,19 +152,17 @@ def in_axes(self, axes: List[str]) -> bool:
...
-
         return {
             "context_id": self.c_id,
             "entity_id": self.entity.identifier,
             "start_date": self.period.start_date,
             "end_date": self.period.end_date,
             "instant": self.period.instant,
+            "filing_name": filing_name,
             **axes_dict
         }
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/xbrl_extract/instance.py]: I've thought about something like this, but the `axes_dict` here is dynamic (will be different for every table)

File: src/ferc_xbrl_extractor/cli.py [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+import sys
+from pathlib import Path
+
+import coloredlogs
+from sqlalchemy import create_engine
+
+from ferc_xbrl_extractor import helpers, xbrl
+from ferc_xbrl_extractor.instance import Instance
+
+DEFAULT_TAXONOMY = "https://eCollection.ferc.gov/taxonomy/form1/2022-01-01/form/form1/form-1_2022-01-01.xsd"
```
Comment (bendnorman) [inline src/ferc_xbrl_extractor/cli.py:14]: So a taxonomy is the schema of an xbrl filing?

File: src/ferc_xbrl_extractor/instance.py [hunk 13]
```diff
@@ -0,0 +1,255 @@
...
+        # Segment node contains dimensions prefixed with xbrldi
+        segment = elem.find(f"{{{XBRL_INSTANCE}}}segment")
+        dims = segment.findall("*") if segment is not None else []
+
+        return cls(
+            identifier=elem.find(f"{{{XBRL_INSTANCE}}}identifier").text,
+            dimensions=[Axis.from_xml(child) for child in dims],
+        )
+
+    def check_dimensions(self, axes: List[str]) -> bool:
```
Comment (bendnorman) [inline src/ferc_xbrl_extractor/instance.py:169]: Should these `check_` methods be pydantic validators?

File: src/ferc_xbrl_extractor/instance.py [hunk 14]
```diff
@@ -0,0 +1,255 @@
...
+            new_context = Context.from_xml(context)
+            context_dict[new_context.c_id] = new_context
+            fact_dict[new_context.c_id] = []
+
+        for fact in facts:
+            new_fact = Fact.from_xml(fact)
+            if new_fact.value is not None:
+                fact_dict[new_fact.c_id].append(new_fact)
+
+        return context_dict, fact_dict, self.name
```
Comment (bendnorman) [inline src/ferc_xbrl_extractor/instance.py:67]: Does it make sense to create a class to hold contexts, dicts and name?

File: src/ferc_xbrl_extractor/datapackage.py [hunk 15]
```diff
@@ -0,0 +1,409 @@
...
+                row.update(contexts[c_id].get_context_ids(filing_name))
+
+                for key, val in row.items():
+                    # Try to parse value from XBRL if not possible it will remain NA
+                    try:
+                        df[key][i] = self.columns[key](val)
+                    except ValueError:
+                        self.logger.warning(
+                            f"Could not parse {val} as {self.columns[key]} in column {key}"
+                        )
```
Comment (bendnorman) [inline src/ferc_xbrl_extractor/datapackage.py:407]: Some comments explaining what each step does here would be helpful.

**Review by bendnorman — COMMENTED** [review]
Amazing work @zschira! I can't believe how complex FERC has made this for us. I left some suggestions and questions for you. Your doc strings and tests gave me a better understanding of XBRL concepts. I think this repo could benefit from some documentation of XBRL concepts with some examples from FERC.

File: setup.py [hunk 16]
```diff
@@ -1,26 +1,107 @@
...
-    entry_points={
-        "console_scripts": [
-            "xbrl_extract = xbrl_extract.cli:main",
-        ]
+    author_email="pudl@catalyst.coop",
+    maintainer="Zach Schira",
+    maintainer_email="zach.schira@catalyst.coop",
+    url="",  # Can be repo or docs URL if no separate web page exists.
+    project_urls={
+        "Source": "https://github.com/catalyst-cooperative/xbrl-extract",
```
[the remark this answers is not in the document]
Comment (zschira) [reply setup.py]: Yes, good catch!

File: src/ferc_xbrl_extractor/helpers.py [hunk 17]
```diff
@@ -0,0 +1,39 @@
...
+
+    Creates an sa.schema.MetaData object reflecting the structure of the
+    database that the passed in ``engine`` refers to, and uses that schema to
+    drop all existing tables.
+
+    Todo:
+        Treat DB connection as a context manager (with/as).
+
+    Args:
+        engine (sa.engine.Engine): An SQL Alchemy SQLite database Engine
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/ferc_xbrl_extractor/helpers.py]: Removed

File: src/ferc_xbrl_extractor/xbrl.py [hunk 18]
```diff
@@ -0,0 +1,164 @@
...
+    taxonomy: str,
+    batch_size: Optional[int] = None,
+    workers: Optional[int] = None,
+    save_metadata: bool = False,
+):
+    """
+    Extract data from all specified XBRL filings.
+
+    Args:
+        instance_paths: List of all XBRL filings to extract.
```
[the remark this answers is not in the document]
Comment (zschira) [reply src/ferc_xbrl_extractor/xbrl.py]: Yeah I've been using the words kind of interchangeably to be honest. I should probably pick one and stick to it

File: src/ferc_xbrl_extractor/instance.py (hunk 13 above)
Comment (zschira) [reply src/ferc_xbrl_extractor/instance.py:169]: These are taking an outside input from a table that is being constructed to see if the context is "valid" for that table

File: src/ferc_xbrl_extractor/instance.py (hunk 14 above)
Comment (zschira) [reply src/ferc_xbrl_extractor/instance.py:67]: Yeah I've been struggling to come up with the best way to handle this. I definitely don't like passing around all these dictionaries though. I think I have a refactor in mind that will clean this up, but I'm leaning towards holding off until after this review is closed. The refactor would touch several parts of the code base, so I don't want it to drag this review out too much, but let me know what you think?

File: src/ferc_xbrl_extractor/datapackage.py (hunk 15 above)
Comment (zschira) [reply src/ferc_xbrl_extractor/datapackage.py]: Added some comments, let me know if they help!

File: src/ferc_xbrl_extractor/cli.py (hunk 12 above)
Comment (zschira) [reply src/ferc_xbrl_extractor/cli.py:14]: Yeah it provides both metadata and defines the structure

File: src/ferc_xbrl_extractor/instance.py (hunk 13 above)
Comment (bendnorman) [reply src/ferc_xbrl_extractor/instance.py:169]: Got it thanks.

File: src/ferc_xbrl_extractor/datapackage.py (hunk 15 above)
Comment (bendnorman) [reply src/ferc_xbrl_extractor/datapackage.py]: Looks good thanks!

Comment (bendnorman) [reply src/ferc_xbrl_extractor/datapackage.py]: Looks good thanks!

**Review by bendnorman — APPROVED** [review]

**Pull request merged**